mc/query/eval.go

```diff
@@ -0,0 +1,372 @@
+	"*":         simpleSelectorAll,
+	"body":      simpleSelectorBody,
+	"id":        simpleSelectorId,
+	"publisher": simpleSelectorPublisher,
+	"namespace": simpleSelectorNamespace,
+	"source":    simpleSelectorSource,
+	"timestamp": simpleSelectorTimestamp}
+
+type FunctionStatementSelector func([]interface{}) []interface{}
+
+func countFunctionSelector(res []interface{}) []interface{} {
+	return []interface{}{len(res)}
+}
+
+var functionSelectors = map[string]FunctionStatementSelector{
+	"COUNT": countFunctionSelector}
+
+func makeResultSet(query *Query) (QueryResultSet, error) {
+	sel := query.selector
+	switch sel := sel.(type) {
+	case SimpleSelector:
+		getf, ok := simpleSelectors[string(sel)]
+		if !ok {
+			return nil, QueryEvalError(fmt.Sprintf("Unexpected selector: %s", sel))
+		}
+
+		return makeSimpleResultSet(getf, query.limit), nil
+
+	case CompoundSelector:
+		getfs := make([]StatementSelector, len(sel))
+		for x, key := range sel {
+			getf, ok := simpleSelectors[string(key)]
+			if !ok {
+				return nil, QueryEvalError(fmt.Sprintf("Unexpected selector: %s", key))
+			}
+			getfs[x] = getf
+		}
+
+		return makeCompoundResultSet(sel, getfs, query.limit), nil
+
+	case *FunctionSelector:
+		fun, ok := functionSelectors[sel.op]
+		if !ok {
+			return nil, QueryEvalError(fmt.Sprintf("Unexpected selector: %s", sel.op))
+		}
+
+		getf, ok := simpleSelectors[string(sel.sel)]
+		if !ok {
+			return nil, QueryEvalError(fmt.Sprintf("Unexpected selector: %s", sel.sel))
+		}
+
+		return makeFunctionResultSet(fun, getf, query.limit), nil
+
+	default:
+		return nil, QueryEvalError(fmt.Sprintf("Unexpected selector type: %T", sel))
+	}
+}
+
+func makeSimpleResultSet(getf StatementSelector, limit int) QueryResultSet {
+	return &SimpleResultSet{rset: make(map[interface{}]bool), getf: getf, limit: limit}
+}
+
+type SimpleResultSet struct {
+	rset  map[interface{}]bool
+	res   []interface{}
+	getf  StatementSelector
+	limit int
+}
+
+func (rs *SimpleResultSet) begin(hint int) {}
+
+func (rs *SimpleResultSet) add(stmt *pb.Statement) {
+	if rs.limit == 0 || len(rs.rset) < rs.limit {
+		rs.rset[rs.getf(stmt)] = true
+	}
+}
+
+func (rs *SimpleResultSet) end() {
+	rs.res = make([]interface{}, len(rs.rset))
+	x := 0
+	for obj, _ := range rs.rset {
+		rs.res[x] = obj
+		x++
+	}
+	rs.rset = nil
+}
+
+func (rs *SimpleResultSet) result() []interface{} {
+	return rs.res
+}
+
+func makeCompoundResultSet(sels []SimpleSelector, getfs []StatementSelector, limit int) QueryResultSet {
+	compf := makeCompoundStatementSelector(sels, getfs)
+	return &CompoundResultSet{getf: compf, limit: limit}
+}
+
+func makeCompoundStatementSelector(sels []SimpleSelector, getfs []StatementSelector) StatementSelector {
+	return func(stmt *pb.Statement) interface{} {
+		val := make(map[string]interface{})
+		for x, sel := range sels {
+			val[string(sel)] = getfs[x](stmt)
+		}
+		return val
+	}
+}
+
+type CompoundResultSet struct {
+	rset  []interface{}
```

Comment:
you mean the different result set semantics?
Well, the simple result set is an actual set, so that there are no duplicates.
The compound result set returns a list of all matches; it's also not possible to implement it with a set because of unhashability.